crates/runtime/src/logic/imports.rs

```diff
@@ -59,6 +60,9 @@ impl VMLogic<'_> {
                 body_len: u64,
                 register_id: u64
             ) -> u32;
+
+            fn generate_uuid(register_id: u64);
```

Comment:
`generate_uuid` can be further generalized to `random_bytes::<N>` that returns random bytes of arbitrary length

Knowledge of the fact that it's a `Uuid` stays with storage, neither the SDK nor runtime need that information, and can be constructed from a `[u8; 16]` via [`uuid::Builder::from_random_bytes`](https://docs.rs/uuid/latest/uuid/struct.Builder.html#method.from_random_bytes)

Reply:
That's an interesting thought. Yes, that could potentially be done, and I can see your reasoning for that. However, in this situation we don't have a need for arbitrary length, and also this would not fit neatly with the UUID functionality. It's definitely a potential for later, but I've left this as-is for now 🙂

Reply:
The concern is, by the time this ships, we'll have a contract to uphold between built developer apps and the interfaces the runtime supports, where there's no reason to.

The ideal approach to designing a bridge is to be liberal within boundaries defined with good reason, and in this case, the generic approach doesn't incur any extra cost.

Randomness has always been a feature we wanted to introduce, it just had to be done safely which is only possible by optimistic execution. As this is the effort that introduces optimistic execution, and a component of its - `uuid` needs randomness, then we have to provide the fundamental facility for randomness.

Otherwise, what's likely to happen after this merges as is, is developers need random numbers and abuse `generate_uuid` to get that. Calling `generate_uuid` N times to get `16N` random bytes and not using it at all for uuids.

Reply:
Hmmmm, I am not sure that there is actually any effect from that, i.e. I do not think the outcome varies.

Let's consider: In a situation where you ask for random bytes, returning `[u8; 16]`, then you get random bytes and can do that as many times as you like. Ergo, the situation you describe as being undesirable. It's the same outcome.

However, in this case we are using semantic identification to recognise that this is a UUID. So it's more constrained than a random number generator of arbitrary length, but also is specific to a purpose of providing UUIDs.

Yes they could call it umpteen times - but they could do this in both cases. What's the difference?

Therefore, on balance, it seems better to retain this as a UUID type, because:

  - There is no difference in randomness or ability to call it multiple times
  - There is semantic identification of the type in use
  - The approach taken to generate a UUID is a specific algorithm and not directly equivalent to an arbitrary randomness generator
  - The exact approach we take to generate UUIDs is abstracted away, and we are free to change it in future without it being tied to arbitrary randomness

Hence it would appear that using a UUID here, i.e. the `Uuid` type, is the correct choice - and even more importantly, that using the UUID functionality for generation and not an arbitrary randomness generator is also correct.

From a contract perspective, the only contract we are formally providing is that an `Id` performs in a particular way. We are free to change `Id` internals later without breaking that contract.

Reply:
> In a situation where you ask for random bytes, returning [u8; 16], then you get random bytes and can do that as many times as you like. Ergo, the situation you describe as being undesirable. It's the same outcome.

as mentioned earlier:

> generate_uuid can be further generalized to random_bytes::<N> that returns random bytes of arbitrary length

it's not the same

> The approach taken to generate a UUID is a specific algorithm and not directly equivalent to an arbitrary randomness generator

there is no "specific algorithm" for UUID v4, the only requirement is you have sufficiently random bytes, and appropriately set the version and variant when making those random bytes into a UUID

> The exact approach we take to generate UUIDs is abstracted away, and we are free to change it in future without it being tied to arbitrary randomness

ideally, if you don't trust a randomness source to be cryptographically secure, you'd pass that as a seed to a CSPRNG like ChaCha. That's what [`StdRng`](https://docs.rs/rand/latest/rand/rngs/struct.StdRng.html) does for example.

Reply:
> That's self-contradicting 🙂 I'm glad you agree with me that v4 UUIDs are not the same as a set of random bytes. As you've described, they have a version and variant built in, which makes them functionally different.

it's not, as mentioned earlier, to construct a v4 uuid with self-provided random bytes, you'd use [`uuid::Builder::from_random_bytes`](https://docs.rs/uuid/latest/uuid/struct.Builder.html#method.from_random_bytes) which proceeds to set the version and variant.

Everything else is random. That's what matters.

Reply:
In order to do that, you still need to expose a source of random bytes, though, even if you are then feeding them into UUID.

Initially, I thought you were talking about returning a `[u8; 16]` of random bytes. I objected to that for obvious reasons.

However, if you are saying you are constructing an actual UUID type then you still need that type available, and you also still need a source of randomness.

I don't see any change to make here - what is already present works, and does what it should, and I'm not sure I even understand fully what you are intending to do, as your objectives are not totally clear to me.

Reply:
Okay, so I think I should recap my previous points:

> Therefore, on balance, it seems better to retain this as a UUID type, because:
> 
>   - There is no difference in randomness or ability to call it multiple times
>   - There is semantic identification of the type in use
>   - The approach taken to generate a UUID is a specific algorithm and not directly equivalent to an arbitrary randomness generator
>   - The exact approach we take to generate UUIDs is abstracted away, and we are free to change it in future without it being tied to arbitrary randomness
> 
> Hence it would appear that using a UUID here, i.e. the Uuid type, is the correct choice - and even more importantly, that using the UUID functionality for generation and not an arbitrary randomness generator is also correct.

I think the best path here is to merge this and then any subsequent changes can be carried out as you see fit - I do not see any harm in UUID being available; the user is able to do the same thing in both cases, and having the actual type present is important and beneficial.